### ISSUE TITLE
Wrong window:close and app:before-quit order of events with autoUpdater.quitAndInstall() 

### ISSUE THREAD

**Issue by salomvary** [electron/electron#7875]
* Electron version: v1.4.5
* Operating system: macOS 10.11.6

Expected:  when the application uses `autoUpdater.quitAndInstall()` to install updates, the `before-quit` event should be triggered before the `close` event on the application's windows. Documentation [confirms this expectation](http://electron.atom.io/docs/api/app/#event-before-quit) and it also is the case for other means of application quit.

Actual: `close` event is triggered first on windows followed by `before-quit` on app.

Example app:

Used the [Quick Start example](http://electron.atom.io/docs/tutorial/quick-start/#write-your-first-electron-app) plus a few lines:

```js
// ...
  win.on('close', (event) => {
    console.log('close')
  })

// ...
app.on('before-quit', function() {
  console.log('before-quit')
})

autoUpdater
  .on('checking-for-update', () => {
    console.log('autoUpdater checking for update')
  })
  .on('update-available', () => {
    console.log('autoUpdater update available')
  })
  .on('update-downloaded', (event, releaseNotes, releaseName) => {
    console.log(`autoUpdater update downloaded ${releaseName}`)
    autoUpdater.quitAndInstall()
  })
  .on('update-not-available', () => {
    console.log('autoUpdater update not available')
  })
  .on('error', error => {
    console.error('autoUpdater error', error)
  })

autoUpdater.setFeedURL(`http://localhost:8080/updates.json?${Math.random()}`)
autoUpdater.checkForUpdates()
``` 

Packaged with this one-liner: `node_modules/.bin/electron-packager . --overwrite --osx-sign.identity="My Developer Identity" && cd your-app-darwin-x64 && zip -r -y your-app-mac.zip your-app.app && cd ..`

Served this `update.json` with `node_modules/.bin/http-server`:
```json
{
    "url": "http://localhost:8080/your-app-darwin-x64/your-app-mac.zip"
}
```

**Comment by zcbenz** [comment]
Due to the API limitation of the auto updater library we used, `autoUpdater.quitAndInstall` currently works by closing all windows first and then try to quit. And it is also too late to change the behavior since apps may already rely on it.

I prefer documenting this explicitly.

**Comment by salomvary** [comment]
Naive question, but why can't the first line of `autoUpdater.quitAndInstall's` implementation be `app.emit('before-quit')`? That would result in less surprises and no need for doing this myself..

If that's not an option, documenting it is much still better than nothing :)

**Comment by MarshallOfSound** [comment]
> Naive question, but why can't the first line of autoUpdater.quitAndInstall's implementation be app.emit('before-quit')? That would result in less surprises and no need for doing this myself..

As @zcbenz it is too late to change this behavior as some apps will have been built with the current order of events in mind.  Changing the order could break existing implementations.

**Comment by salomvary** [comment]
> Changing the order could break existing implementations.

That's true. So maybe in the next major version release? ;)

**Comment by kevinsawicki** [comment]
Fixed by #7986
